python/ray/rllib/env/external_multi_agent_env.py

```diff
@@ -0,0 +1,262 @@
+            observation (dict): Current environment observation.
+            action (dict): Action for the observation.
+        """
+
+        # FIXME handle different agents here
+        episode = self._get(episode_id)
+        episode.log_action(observation, action)
+
+    @PublicAPI
+    def log_returns(self, episode_id, reward, info=None):
+        """Record returns from the environment.
+
+        The reward will be attributed to the previous action taken by the
+        episode. Rewards accumulate until the next action. If no reward is
+        logged before the next action, a reward of 0.0 is assumed.
+
+        Arguments:
+            episode_id (str): Episode id returned from start_episode().
+            reward (dict): Reward from the environment agents.
+            info (dict): Optional info dict.
+        """
+
+        episode = self._get(episode_id)
+
+        # accumulate reward by agent
+        # for existing agents, we want to add the reward up
+        for agent, rew in reward.items():
+            if agent in episode.cur_reward:
+                episode.cur_reward[agent] += rew
+            else:
+                episode.cur_reward[agent] = rew 
+        if info:
+            episode.cur_info = info or {}
+
+    @PublicAPI
+    def end_episode(self, episode_id, observation):
+        """Record the end of an episode.
+
+        Arguments:
+            episode_id (str): Episode id returned from start_episode().
+            observation (obj): Current environment observation.
+        """
+
+        episode = self._get(episode_id)
+        self._finished.add(episode.episode_id)
+        episode.done(observation)
+
+    def _get(self, episode_id):
+        """Get a started episode or raise an error."""
+
+        if episode_id in self._finished:
+            raise ValueError(
+                "Episode {} has already completed.".format(episode_id))
+
+        if episode_id not in self._episodes:
+            raise ValueError("Episode {} not found.".format(episode_id))
+
+        return self._episodes[episode_id]
+
+
+class _ExternalEnvEpisode(object):
+    """
+    Tracked state for each active episode.
+    FIXME fix this class:
+    - multiagent dict variables
+    """
+
+    def reset_cur_done(self, done=False):
+        self.cur_done = {"__all__": done}
+
+    def __init__(self, episode_id, results_avail_condition, training_enabled):
+        self.episode_id = episode_id
+        self.results_avail_condition = results_avail_condition
+        self.training_enabled = training_enabled
+        self.data_queue = queue.Queue()
+        self.action_queue = queue.Queue()
+        self.new_observation = None
+        self.new_action = None
+
+        # FIXME dirty. do we need to know all agents here?
+        # if I do not set this, the first _send is going to fail
+        # and result in a keyerror in the evaluator
+        self.cur_reward = { x: 0.0 for x in list(range(200)) } 
```

Comment:
Hm we can probably fix the evaluator to not raise an error. What's the full stack trace?

Or, you can track the agent ids that have appeared so far and supply them here.

Reply:
The full stack trace is:
```
Traceback (most recent call last):
  File "cartpole_server.py", line 126, in <module>
    print(pretty_print(dqn_trainer.train()))
  File "/usr/lib/python3.7/site-packages/ray/rllib/agents/agent.py", line 293, in train
    result = Trainable.train(self)
  File "/usr/lib/python3.7/site-packages/ray/tune/trainable.py", line 150, in train
    result = self._train()
  File "/usr/lib/python3.7/site-packages/ray/rllib/agents/dqn/dqn.py", line 258, in _train
    self.optimizer.step()
  File "/usr/lib/python3.7/site-packages/ray/rllib/optimizers/sync_replay_optimizer.py", line 87, in step
    batch = self.local_evaluator.sample()
  File "/usr/lib/python3.7/site-packages/ray/rllib/evaluation/policy_evaluator.py", line 391, in sample
    batches = [self.input_reader.next()]
  File "/usr/lib/python3.7/site-packages/ray/rllib/evaluation/sampler.py", line 41, in next
    batches = [self.get_data()]
  File "/usr/lib/python3.7/site-packages/ray/rllib/evaluation/sampler.py", line 80, in get_data
    item = next(self.rollout_provider)
  File "/usr/lib/python3.7/site-packages/ray/rllib/evaluation/sampler.py", line 283, in _env_runner
    preprocessors, obs_filters, unroll_length, pack, callbacks)
  File "/usr/lib/python3.7/site-packages/ray/rllib/evaluation/sampler.py", line 371, in _process_observations
    rewards[env_id][agent_id] or 0.0))
KeyError: 0
```
In the example, this happens in the first ``client.get_action(...)`` call.
As this is the first call containing keyed agent information , we could possibly supply all possible agents earlier, e.g. in ``client.start_episode(...)``?

Reply:
Fixed.